cost_reports/finops.py

```diff
@@ -0,0 +1,184 @@
+    'No Program / 000000':'NO PROGRAM / 000000'
+}
+
+
+def main(cur_parquet, account_csv):
+    report_body = ""
+
+    # Read in the account labels, for those accounts whose costs are all allocated the same way
+    accounts_columns = {
+        'account_id':'str',
+        ACCOUNT_NAME_COLUMN:'str',
+        ACCOUNT_COST_CENTER_COLUMN:'str'
+    }
+    accounts = pd.read_csv(account_csv, usecols=accounts_columns.keys(), dtype=accounts_columns)
+
+    # Read in the month's cost and usage report
+    columns=[
+        VALUE_COLUMN,
+        ACCOUNT_ID_COLUMN,
+        COST_CENTER_COLUMN,
+        COST_CENTER_OTHER_COLUMN,
+        CLOUDFORMATION_STACK_COLUMN,
+        SERVICE_CATALOG_PRINCIPAL_COLUMN,
+        LINE_ITEM_TYPE_COLUMN,
+        PRODUCT_NAME_COLUMN,
+        USAGE_TYPE_COLUMN,
+        DESCRIPTION_COLUMN
+    ]
+
+    acct_rows = pd.read_parquet(cur_parquet, columns=columns)
+
+    # Compute total cost for the month
+    total_cost = acct_rows[VALUE_COLUMN].sum()
+
+    # we now want to inner join df with accounts on ACCOUNT_ID_COLUMN <-> 'account_id'
+    accounts = accounts.rename(columns={'account_id': ACCOUNT_ID_COLUMN})
+    joined = acct_rows.join(accounts.set_index(ACCOUNT_ID_COLUMN), on=ACCOUNT_ID_COLUMN, how="inner")
+    report_body += f"Total line items: {joined.shape[0]}\n"
+
+    # either cost center is not null or account cost center is not null (one of the two IS set)
+    has_center = joined[
+        (
+            (joined[COST_CENTER_COLUMN].notna()) &
+            (joined[COST_CENTER_COLUMN] != 'na')
+        ) | (
+            (joined[ACCOUNT_COST_CENTER_COLUMN].notna()) &
+            (joined[ACCOUNT_COST_CENTER_COLUMN] != 'na')
+        )
+    ]
+    report_body += f"Rows with Cost Center data: {has_center.shape[0]}\n"
+
+    # cost center is either null or 'na' AND account cost center is either null or 'na'
+    no_center = joined[
+        (
+            (joined[COST_CENTER_COLUMN].isna()) |
+            (joined[COST_CENTER_COLUMN] == 'na')
+        ) & (
+            (joined[ACCOUNT_COST_CENTER_COLUMN].isna()) |
+            (joined[ACCOUNT_COST_CENTER_COLUMN] == 'na')
+        )
+    ]
+    report_body += f"Rows without Cost Center data: {no_center.shape[0]}\n"
+
+    # group expenses by their cost center.  See group_by() for details
+    by_center = acct_rows.groupby((lambda row_index: group_by_cost_center(joined,row_index)), dropna=False, group_keys=False)
+
+    # group expenses by their cloudformation stack.  See group_by() for details
+    by_stack = no_center.groupby((lambda row_index: group_by_tag(no_center,row_index,CLOUDFORMATION_STACK_COLUMN,"Cloudformation Stack")), dropna=False, group_keys=False)
+
+    # group expenses by their service catalog provisioner.  See group_by() for details
+    by_catalog = no_center.groupby((lambda row_index: group_by_tag(no_center,row_index,SERVICE_CATALOG_PRINCIPAL_COLUMN,"Service Catalog Provisioner")), dropna=False, group_keys=False)
```

Comment:
What is the purpose of grouping by the provisioning principal?  Is it part of the effort to group by user?  If so, why not add in the per-user costs from other accounts?  Moreover in Service Catalog (scipool-prod) the provisioned products have a tag which is the Synapse ID of the user, and which is probably a better identifier for the individual).

Reply:
This was an initial attempt to discover information about rows missing tag information (and more generally to understand the data being processed), but the effort to analyze untracked costs should probably be separate from this ETL process.